### PR TITLE
Extracted coffee core

### LINKED ISSUE #61: extract the coffe plugin manager from the coffe_cmd and move in a new crate
# Background

To run Coffe you are run the following command `cargo run -- --help` and you can run the following command to setup coffee for your core lightning node

```
cargo run -- setup <path of the core lightning file>
cargo run -- remote add lightningd https://github.com/lightningd/plugins.git
cargo run -- install <plugin_name>
```

However in order to implement https://github.com/coffee-tools/coffee/issues/60 and https://github.com/coffee-tools/coffee/issues/59 without code duplication we should move the coffee plugin manager implementation inside a new crater `coffee_core` to allow different kind of interface such as plugins and or web API.

# Goal

The goal is to create a new rust crate with the name `coffee_core` and move the implementation of coffee inside this new crate. Then update the `coffee_cmd` to use this new crate

### PR DESCRIPTION
**Pull request by tareknaser360**
Fixes #61

### CODE REVIEW

File: coffee_core/Cargo.toml [hunk 1]
```diff
@@ -0,0 +1,19 @@
+[package]
+name = "coffee_core"
+version = "0.0.1-alpha.1"
+edition = "2021"
+
+
+[dependencies]
+tokio = { version = "1.22.0", features = ["full"] }
+clap = { version = "4.0.26", features = ["derive"] }
```
Comment (vincenzopalazzo) [inline coffee_core/Cargo.toml:9]:
```suggestion
clap = { version = "4.0.26", features = ["derive"] }
```
this is needed?

File: coffee_core/src/coffee.rs [hunk 2]
```diff
@@ -22,8 +22,8 @@ use serde_json::Value;
 use std::fmt::Debug;
 use std::vec::Vec;
 
-pub mod cmd;
-mod config;
+//use super::cmd;
```
Comment (vincenzopalazzo) [inline coffee_core/src/coffee.rs:25]:
```suggestion
```

**Review by vincenzopalazzo — CHANGES_REQUESTED** [review]
Ok now that I'm thinking about it, I think the coffee_core/src/cmd.rs should be staying inside the cmd crate, because it is a command-line dependency.

We do not have the command line in a web server or a core lightning plugin

**Comment by tareknaser360** [comment]
> Ok now that I'm thinking about it, I think the coffee_core/src/cmd.rs should be staying inside the cmd crate, because it is a command-line dependency.
> 
> We do not have the command line in a web server or a core lightning plugin

- `coffee_core/src/coffee.rs` uses functions from `coffee_core/src/cmd.rs` 
- `coffee_cmd/src/main.rs` uses functions from `coffee_core/src/coffee.rs` 

So if we move `cmd.rs` back to the cmd crate we will have cyclic dependencies. 
Is there any workaround?

**Comment by vincenzopalazzo** [comment]
>So if we move cmd.rs back to the cmd crate we will have cyclic dependencies.
We are in rust I think cyclic dependencies are not a problem, but after your patch, I will try to remove the cmd dependencies from the core

**Comment by vincenzopalazzo** [comment]
>Sorry, I can't get it to work with cyclic dependencies.

Oh, this is my bad, yes please push the code and I will provide a solution with a diff for it :)

**Comment by tareknaser360** [comment]
Thanks for your help.

What is there left to do?

**Comment by vincenzopalazzo** [comment]
>What is there left to do?

Rebase and make the CI happy again, after that I do not know I think there is all that we need? make a double  if I' am forgetting something

**Comment by tareknaser360** [comment]
I think now the core functions are in a separate crate that we can use later in the server.

**Comment by tareknaser360** [comment]
Some docs checks are neutral. Is this expected? There were some merge conflicts with the rebase

File: docs/index.mdx [hunk 3]
```diff
@@ -1,42 +0,0 @@
-# Coffee Plugin Manager
-
-Welcome to **Coffee**, the plugin manager for Core Lightning. 
-It takes care of all the configuration and installation of a plugin for your Core Lightning node. 
-
-Using Coffee, anyone can manage plugins for their Core Lightning node. 
-Coffee handles all the tedious setup and configuration tasks of the underlying 
-Core Lightning plugin infrastructure and the hassle of setting up dependencies.
```
Comment (vincenzopalazzo) [inline docs/index.mdx:8]: this needs to stay, it is used to build the https://docs.page/coffee-tools/coffee

File: docs/index.mdx [hunk 4]
```diff
@@ -1,42 +0,0 @@
...
-Coffee offers a range of resources and libraries to simplify integrating lightning plugins into 
-your business applications :smile:.
-
-
-| Package        | Description                                                     | Version    |
-|----------------|-----------------------------------------------------------------|------------|
-| [coffee_cmd](../coffee_cmd/)     | Package providing CLI to the Coffee plugin manager      | pre_release |
-| [coffee_github](../coffee_github/)     | GitHub interface to the Coffee plugin manager      | pre_release |
-| [coffee_lib](../coffee_lib/)     | The core library to the Coffee plugin ecosystem      | pre_release |
-| [coffee_storage](../coffee_storage/)     | The local storage model package for the Coffee plugin manager     | pre_release |
```
Comment (vincenzopalazzo) [inline docs/index.mdx:27]: you need to add the coffee_core in the table and also in the Readme.

In addition, you need to add the `core` in the scope https://coffee-docs.netlify.app/contributing-to-coffee.html#scopes

Docs is painful :)

**Review by vincenzopalazzo — CHANGES_REQUESTED** [review]
Looks almost good!

File: docs/index.mdx (hunk 3 above)
Comment (tareknaser360) [reply docs/index.mdx]: Yes. I didn't notice

**Review by vincenzopalazzo — APPROVED** [review]
Thanks!

From the next time can you follow the following https://coffee-docs.netlify.app/contributing-to-coffee.html guideline?

Thanks

**Comment by tareknaser360** [comment]
Yes sure

**Pull request merged**